// hikerJoy/config.js

var env = 'dev';

var dbCols = {
    'users': 'users',
    'orgs': 'organizations',
    'userMsg': 'userMessages',
    'pwdReset': 'passwordReset',
    'acts': 'activities',
    'userActs': 'userActs',
    'emails': 'emails',
    'actTags': 'activityTags',
    'weixin': 'weixin',
    'sessions': 'sessions',
    'forumPost': 'forumPosts',
    'forumPostPrm': 'forumPostPrms',
    'forumPostSec': 'forumPostSecs',
    'cacheInfo': 'cacheInfo',
    'log': {
        'critical': 'log_critical',
        'error': 'log_error',
        'warning': 'log_warning',
        'info': 'log_info',
        'debug': 'log_debug',
        'update': 'log_update',
        'page': 'log_page' //by dataServer
    }
};

var ports = {
    'fileServer': 85,
    'dataServer': 1337,
    'logServer': 1338,
    'authServer': 1339,
    'fileUpload': 1340,
    'weixinServer': 1341
};

var cachedInfo = {
    'orgContextUpdatedOn': 'orgContextUpdatedOn'
};

var orgWeixinKeywords = ['activities', 'intro'];


// if (env === 'dev') {
    // ports.fileServer = 85;
    // exports.dbName = 'hikerJoy';
    // exports.siteUrl = ports.fileServer === 80 ? 'http://127.0.0.1' : 'http://127.0.0.1:' + ports.fileServer;
    // exports.siteDomain = '127.0.0.1';
    // exports.forumPostSecColdDownInSeconds = 0;
    // exports.forumPostPrmColdDownInSeconds = 0;
    // exports.forumPostColdDownInSeconds = 0;
    // exports.staticFile304 = false;
    // exports.cacheStaticFile = false;
    // exports.cachePageJade = false;
    // exports.canSendEmail = false;
    // exports.cacheEmailTemplates = false;
// }
// else if (env === 'prod') {
    ports.fileServer = 80;
    exports.dbName = 'hikerJoy';
    exports.siteUrl = ports.fileServer === 80 ? 'http://www.hikerjoy.com' : 'http://www.hikerjoy.com:' + ports.fileServer;
    exports.siteDomain = 'www.hikerjoy.com';
    exports.forumPostSecColdDownInSeconds = 5;  //5 secs
    exports.forumPostPrmColdDownInSeconds = 30;  //30 secs
    exports.forumPostColdDownInSeconds = 60 * 5;  //5 minute
    exports.staticFile304 = true;
    exports.cacheStaticFile = true;
    exports.cachePageJade = true;
    exports.canSendEmail = true;
    exports.cacheEmailTemplates = true;
// }
// else { //test
    // ports.fileServer = 80;
    // exports.dbName = 'hikerJoy';
    // exports.siteUrl = ports.fileServer === 80 ? 'http://hikerjoytest.cloudapp.net' : 'http://hikerjoytest.cloudapp.net:' + ports.fileServer;
    // exports.siteDomain = 'hikerjoytest.cloudapp.net';
    // exports.forumPostSecColdDownInSeconds = 0;
    // exports.forumPostPrmColdDownInSeconds = 0;
    // exports.forumPostColdDownInSeconds = 0;
    // exports.staticFile304 = true;
    // exports.cacheStaticFile = true;
    // exports.cachePageJade = true;
    // exports.canSendEmail = false;
    // exports.cacheEmailTemplates = true;
// }


exports.ports = ports;
exports.sessionExpMin = 60 * 24;
exports.tokenExpMin = 60 * 24;
exports.logDbName = 'hikerJoy_log';
exports.dbCols = dbCols;
exports.smtpHost = 'smtp.163.com';
exports.smtpPort = '25';
exports.smtpAuth = {
    'user': 'hikerjoy@163.com',
    'pass': 'cy159357CY'
};
exports.smtpFrom = "hikerJoy<hikerjoy@163.com>";
exports.defaultForumPostPageSize = 30;
exports.defaultForumPostPrmPageSize = 10;
exports.defaultForumPostSecPageSize = 5;
exports.minForumPostPageSize = 15;
exports.minForumPostPrmPageSize = 5;
exports.minForumPostSecPageSize = 5;
exports.maxForumPostPageSize = 100;
exports.maxForumPostPrmPageSize = 30;
exports.maxForumPostSecPageSize = 10;
exports.systemStartsOn = new Date();
exports.doLog = true;
exports.cachedInfo = cachedInfo;
exports.orgWeixinKeywords = orgWeixinKeywords;
