// hikerJoy/constants.js

var orgStatus = {
    'active': 10,
    'inactive': 20
};

var activityStatus = {
    'prepare': 110,
    'open': 120,
    'going': 130,
    'closed': 140,
    'archived': 150,
    'removed': 160
};

//active activity means: not archived, not removed
var activeActStatus = [activityStatus.prepare, activityStatus.open, activityStatus.going, activityStatus.closed];

//active activity means: closed, or archived
var historicalActStatus = [activityStatus.closed, activityStatus.archived];

var userStatus = {
    'active': 210,
    'inactive': 220
};

var actMemberStatus = {
    'queued': 410,
    'inTeam': 420,
    'kickout': 430,
    'quit': 440
};

var actMemberStatusNames = {
    '410': '等待',
    '420': '入队',
    '430': '强制退出',
    '440': '退出'
};

var activeActMemberStatus = [actMemberStatus.queued, actMemberStatus.inTeam];
var inactiveActMemberStatus = [actMemberStatus.kickout, actMemberStatus.quit];
var organizerUpdateMemberStatus = [actMemberStatus.kickout, actMemberStatus.queued, actMemberStatus.inTeam];

var emailStatus = {
    'send': 510,
    'failed': 520
};

var weixinReplyStatus = {
    'active': 610,
    'inactive': 620
};

var forumPostStatus = {
    'active': 710,
    'inactive': 720
};

var forumPostPrmStatus = {
    'active': 810,
    'inactive': 820
};

var forumPostSecStatus = {
    'active': 910,
    'inactive': 920
};

var logLevel = {
    'critical': 0,
    'error': 1,
    'warning': 2,
    'info': 3,
    'debug': 4
};

var register_rc = {
    'success': { 'rc': 0, 'msg': '注册成功！' },
    'emailAlreadyExists': { 'rc': 1, 'msg': '注册失败：邮箱地址已存在，请尝试找回密码，或者使用另一个邮箱地址 ^^' },
    'inputParaError': { 'rc': 2, 'msg': '注册失败：输入参数错误' },
    'unKnownError': { 'rc': 3, 'msg': '注册失败：未知错误' },
    'nickNameDup': { 'rc': 4, 'msg': '注册失败：昵称已被占用，请换一个吧 :(' }
};

var changePwd_rc = {
    'success': { 'rc': 0, 'msg': '修改密码成功！' },
    'userNotFound': { 'rc': 1, 'msg': '修改密码失败：请先登录，并再次尝试' },
    'pwdNotMatch': { 'rc': 2, 'msg': '修改密码失败：原密码不匹配' },
    'inputParaError': { 'rc': 3, 'msg': '修改密码失败：输入参数错误' },
    'unKnownError': { 'rc': 4, 'msg': '修改密码失败：未知错误' }
};

var updateUserEmail_rc = {
    'success': { 'rc': 0, 'msg': '修改登录邮箱成功！' },
    'userNotFound': { 'rc': 1, 'msg': '修改登录邮箱失败：请先登录，并再次尝试' },
    'noChange': { 'rc': 2, 'msg': '修改登录邮箱失败：用户已经是指定登录邮箱，无需修改' },
    'emailDuplicate': { 'rc': 3, 'msg': '修改登录邮箱失败：指定邮箱已存在' },
    'pwdNotMatch': { 'rc': 4, 'msg': '修改登录邮箱失败：密码不匹配' },
    'inputParaError': { 'rc': 5, 'msg': '修改登录邮箱失败：输入参数错误' },
    'unKnownError': { 'rc': 6, 'msg': '修改登录邮箱失败：未知错误' }
};

var sendUserMessage_rc = {
    'success': { 'rc': 0, 'msg': '发送消息成功！' },
    'senderNotFound': { 'rc': 1, 'msg': '发送消息失败：请先登录，并再次尝试' },
    'receiverNotFound': { 'rc': 2, 'msg': '发送消息失败：没有找到收信人 TAT' },
    'noSendToSelf': { 'rc': 3, 'msg': '发送消息失败：不能发送给自己' },
    'senderNoNickName': { 'rc': 4, 'msg': '发送消息失败：请先设置自己的昵称，再尝试发送 ^^' },
    'inputParaError': { 'rc': 5, 'msg': '发送消息失败：输入参数错误' },
    'unKnownError': { 'rc': 6, 'msg': '发送消息失败：未知错误' }
};

var updateOrgPic_rc = {
    'success': { 'rc': 0, 'msg': '更新组织图片成功！' },
    'uploadImageFail': { 'rc': 1, 'msg': '更新组织图片失败：图片上传失败，请联系系统管理员' },
    'notAuth': { 'rc': 2, 'msg': '更新组织图片失败：未授权，请登录并再次尝试，仍然不行请联系系统管理员' },
    'inputParaError': { 'rc': 3, 'msg': '更新组织图片失败：输入参数错误' },
    'unKnownError': { 'rc': 4, 'msg': '更新组织图片失败：未知错误' }
};

var defaultPwd = {
    'plain': 'China2014',
    'hash': '2aac27b9701784549797ef6675835860',
}

var claimPwdReset_rc = {
    'success': { 'rc': 0, 'msg': '重置密码成功！请使用：' + defaultPwd.plain + ' 登录，并尽快修改密码' },
    'fail': { 'rc': 1, 'msg': '重置密码失败：申请不存在或已过期，请重新尝试' },
    'unKnownError': { 'rc': 2, 'msg': '重置密码失败：未知错误' }
};

var pwdResetRequest_rc = {
    'success': { 'rc': 0, 'msg': '申请密码重置成功！请查收邮件' },
    'inputParaError': { 'rc': 1, 'msg': '申请密码重置失败：输入参数错误' },
    'userNotFound': { 'rc': 2, 'msg': '申请密码重置失败：没有找到对应用户' },
    'activeRequest': { 'rc': 3, 'msg': '申请密码重置失败：该用户已经申请密码重置，请查收邮件' },
    'unKnownError': { 'rc': 4, 'msg': '申请密码重置失败：未知错误' }
};

var upsertOrgSignupSheetTemplate_rc = {
    'success': { 'rc': 0, 'msg': '更新活动模板成功！' },
    'inputParaError': { 'rc': 1, 'msg': '更新活动模板失败：输入参数错误，请确认提交模板含有内容' },
    'inputParaInvalid': { 'rc': 2, 'msg': '更新活动模板失败：输入参数未通过验证！' }, //for debug
    'notAuth': { 'rc': 3, 'msg': '更新活动模板失败：未授权' },
    'unKnownError': { 'rc': 4, 'msg': '更新活动模板失败：未知错误' }
};

var renameOrgSignupSheetTemplate_rc = {
    'success': { 'rc': 0, 'msg': '重命名活动模板成功！' },
    'inputParaError': { 'rc': 1, 'msg': '重命名活动模板失败：输入参数错误，请确认新名称' },
    'templateNotFound': { 'rc': 2, 'msg': '重命名活动模板失败：未找到指定重命名的模板！' },
    'notAuth': { 'rc': 3, 'msg': '重命名活动模板失败：未授权' },
    'nameNoChange': { 'rc': 4, 'msg': '重命名活动模板失败：新名称与原先一致' },
    'duplicateName': { 'rc': 5, 'msg': '重命名活动模板失败：存在名称相同的模板，请换一个重试' },
    'unKnownError': { 'rc': 6, 'msg': '重命名活动模板失败：未知错误' }
};

var deleteOrgSignupSheetTemplate_rc = {
    'success': { 'rc': 0, 'msg': '删除活动模板成功！' },
    'inputParaError': { 'rc': 1, 'msg': '删除活动模板失败：输入参数错误，请确认提交模板含有内容' },
    'templateNotFound': { 'rc': 2, 'msg': '删除活动模板失败：未找到指定删除的模板！' },
    'notAuth': { 'rc': 3, 'msg': '删除活动模板失败：未授权' },
    'unKnownError': { 'rc': 4, 'msg': '删除活动模板失败：未知错误，请联系系统管理员！' }
};

var submitOrgActivity_rc = {
    'success': { 'rc': 0, 'msg': '提交活动成功！' },
    'inputParaError': { 'rc': 1, 'msg': '提交活动失败：输入参数错误' },
    'orgIdNotMatch': { 'rc': 2, 'msg': '提交活动失败：不能修改活动所属的组织！' },
    'notAuth': { 'rc': 3, 'msg': '提交活动失败：未授权' },
    'uploadImageFail': { 'rc': 4, 'msg': '提交活动失败：上传图片失败' },
    'actNotFound': { 'rc': 5, 'msg': '提交活动失败：未找到要求修改的组织活动' },
    'noOrganizer': { 'rc': 6, 'msg': '提交活动失败：未指定活动领队，请至少指定一名' },
    'organizerNotFound': { 'rc': 7, 'msg': '提交活动失败：至少一名活动领队不存在或不具有组织活动的权限' },
    'unKnownError': { 'rc': 8, 'msg': '提交活动失败：未知错误，请联系系统管理员！' }
};

var updateUserPersonalInfo_rc = {
    'success': { 'rc': 0, 'msg': '更新个人信息成功！' },
    'inputParaError': { 'rc': 1, 'msg': '更新个人信息失败：输入参数错误' },
    'userNotFound': { 'rc': 2, 'msg': '更新个人信息失败：未找到用户' },
    'unKnownError': { 'rc': 3, 'msg': '更新个人信息失败：未知错误' },
    'nickNameDup': { 'rc': 4, 'msg': '更新个人信息失败：昵称已被占用，请换一个吧 :(' }
};

var signUp_rc = {
    'success': { 'rc': 0, 'msg': '报名成功！' },
    'successWithTempUser': { 'rc': 1, 'msg': '报名成功！已为你创建用户，密码：' + defaultPwd.plain + '，请尽快修改密码！' },
    'inputParaError': { 'rc': 2, 'msg': '报名失败：输入参数错误' },
    'userNotFound': { 'rc': 3, 'msg': '报名失败：未找到当前用户' },
    'unKnownError': { 'rc': 4, 'msg': '报名失败：未知错误' },
    'tmpUserEmailExists': { 'rc': 5, 'msg': '报名失败：创建临时用户失败，邮箱已存在，请换一个并重试 ^^' },
    'actNotFound': { 'rc': 6, 'msg': '报名失败：未找到指定的活动' },
    'userInAnotherAct': { 'rc': 7, 'msg': '报名失败：检测到已报名另一个相同时间的活动，请先退出，并再次尝试' },
    'special': { 'rc': 8, 'msg': '报名失败：特殊身份账号不能参与活动报名' }
};

var userActQuit_rc = {
    'success': { 'rc': 0, 'msg': '活动退出成功！' },
    'userNotFound': { 'rc': 1, 'msg': '活动退出失败：未找到当前用户' },
    'actNotFound': { 'rc': 2, 'msg': '活动退出失败：指定退出的活动未找到' },
    'memStatusNotAllowQuit': { 'rc': 3, 'msg': '活动退出失败：当前你的状态不允许退出' },
    'actStatusNotAllowQuit': { 'rc': 4, 'msg': '活动退出失败：当前活动的状态不允许退出' },
    'unKnownError': { 'rc': 5, 'msg': '活动退出失败：未知错误' }
};

var updateUserSignupSheet_rc = {
    'success': { 'rc': 0, 'msg': '更新报名信息成功！' },
    'userNotFound': { 'rc': 1, 'msg': '更新报名信息失败：未找到当前用户' },
    'actNotFound': { 'rc': 2, 'msg': '更新报名信息失败：指定退出的活动未找到' },
    'memStatusNotAllowUpdate': { 'rc': 3, 'msg': '更新报名信息失败：当前你的状态不允许修改报名信息' },
    'actStatusNotAllowUpdate': { 'rc': 4, 'msg': '更新报名信息失败：当前活动的状态不允许修改报名信息' },
    'unKnownError': { 'rc': 5, 'msg': '更新报名信息失败：未知错误' },
    'inputParaError': { 'rc': 6, 'msg': '更新报名信息失败：输入参数错误' }
};

var grantUserOrgAdmin_rc = {
    'success': { 'rc': 0, 'msg': '管理员授权成功！' },
    'notAuth': { 'rc': 1, 'msg': '管理员授权失败：未授权' },
    'alreadyAdmin': { 'rc': 2, 'msg': '管理员授权失败：此用户已经是管理员' },
    'userNotFound': { 'rc': 3, 'msg': '管理员授权失败：指定用户不存在' },
    'cannotAuthorize': { 'rc': 4, 'msg': '管理员授权失败：无法对指定用户进行管理员授权' },
    'unKnownError': { 'rc': 5, 'msg': '管理员授权失败：未知错误' },
    'inputParaError': { 'rc': 6, 'msg': '管理员授权失败：输入参数错误' }
};

var setOrgActArchived_rc = {
    'success': { 'rc': 0, 'msg': '活动归档成功！' },
    'notAuth': { 'rc': 1, 'msg': '活动归档失败：未授权' },
    'actNotFound': { 'rc': 2, 'msg': '活动归档失败：指定归档的活动未找到' },
    'archived': { 'rc': 3, 'msg': '活动归档失败：指定活动已经归档' },
    'inputParaError': { 'rc': 4, 'msg': '活动归档失败：输入参数错误' },
    'unKnownError': { 'rc': 5, 'msg': '活动归档失败：未知错误' }
};

var setOrgActRemoved_rc = {
    'success': { 'rc': 0, 'msg': '活动删除成功！' },
    'notAuth': { 'rc': 1, 'msg': '活动删除失败：未授权' },
    'actNotFound': { 'rc': 2, 'msg': '活动删除失败：指定删除的活动未找到' },
    'inputParaError': { 'rc': 3, 'msg': '活动删除失败：输入参数错误' },
    'unKnownError': { 'rc': 4, 'msg': '活动删除失败：未知错误' }
};

var removeOrgAdmin_rc = {
    'success': { 'rc': 0, 'msg': '移除组织管理员成功！' },
    'notAuth': { 'rc': 1, 'msg': '移除组织管理员失败：未授权' },
    'userNotFound': { 'rc': 2, 'msg': '移除组织管理员失败：未找到指定用户' },
    'notAdmin': { 'rc': 3, 'msg': '移除组织管理员失败：指定用户不是组织管理员' },
    'inputParaError': { 'rc': 4, 'msg': '移除组织管理员失败：输入参数错误' },
    'unKnownError': { 'rc': 5, 'msg': '移除组织管理员失败：未知错误' },
    'orgNotFound': { 'rc': 6, 'msg': '移除组织管理员失败：未找到指定组织' }
};

var updateOrgIntro_rc = {
    'success': { 'rc': 0, 'msg': '保存组织介绍成功！' },
    'notAuth': { 'rc': 1, 'msg': '保存组织介绍失败：未授权' },
    'orgNotFound': { 'rc': 2, 'msg': '保存组织介绍失败：未找到指定组织' },
    'inputParaError': { 'rc': 3, 'msg': '保存组织介绍失败：输入参数错误' },
    'unKnownError': { 'rc': 4, 'msg': '保存组织介绍失败：未知错误' }
};

var updateMemberStatus_rc = {
    'success': { 'rc': 0, 'msg': '更新队员状态成功！' },
    'notAuth': { 'rc': 1, 'msg': '更新队员状态失败：未授权' },
    'memberNotFound': { 'rc': 2, 'msg': '更新队员状态失败：未找到指定队员' },
    'failedMemberStatus': { 'rc': 3, 'msg': '更新队员状态失败：此队员的当前状态不允许其进行状态更新' },
    'failedActivityStatus': { 'rc': 4, 'msg': '更新队员状态失败：此队员所属活动的状态不允许其进行状态更新' },
    'inputParaError': { 'rc': 5, 'msg': '更新队员状态失败：输入参数错误' },
    'unKnownError': { 'rc': 6, 'msg': '更新队员状态失败：未知错误' }
};

var reactivateOrg_rc = {
    'success': { 'rc': 0, 'msg': '重新激活组织成功！' },
    'notAuth': { 'rc': 1, 'msg': '重新激活组织失败：未授权' },
    'orgNotFound': { 'rc': 2, 'msg': '重新激活组织失败：未找到指定组织' },
    'active': { 'rc': 3, 'msg': '重新激活组织失败：该组织已经是激活状态' },
    'inputParaError': { 'rc': 4, 'msg': '重新激活组织失败：输入参数错误' },
    'unKnownError': { 'rc': 5, 'msg': '重新激活组织失败：未知错误' }
};

var disactivateOrg_rc = {
    'success': { 'rc': 0, 'msg': '停运组织成功！' },
    'notAuth': { 'rc': 1, 'msg': '停运组织失败：未授权' },
    'orgNotFound': { 'rc': 2, 'msg': '停运组织失败：未找到指定组织' },
    'inactive': { 'rc': 3, 'msg': '停运组织失败：该组织已经是停运状态' },
    'inputParaError': { 'rc': 4, 'msg': '停运组织失败：输入参数错误' },
    'unKnownError': { 'rc': 5, 'msg': '停运组织失败：未知错误' }
};

var updateOrgBasicInfo_rc = {
    'success': { 'rc': 0, 'msg': '更新组织信息成功！' },
    'notAuth': { 'rc': 1, 'msg': '更新组织信息失败：未授权' },
    'orgNotFound': { 'rc': 2, 'msg': '更新组织信息失败：未找到指定组织' },
    'inputParaError': { 'rc': 3, 'msg': '更新组织信息失败：输入参数错误' },
    'aliasDup': { 'rc': 4, 'msg': '更新组织信息失败：检测到其Alias与另一个组织重复' },
    'unKnownError': { 'rc': 5, 'msg': '更新组织信息失败：未知错误' }
};

var createNewOrg_rc = {
    'success': { 'rc': 0, 'msg': '添加组织成功！' },
    'notAuth': { 'rc': 1, 'msg': '添加组织失败：未授权' },
    'inputParaError': { 'rc': 2, 'msg': '添加组织失败：输入参数错误' },
    'aliasDup': { 'rc': 3, 'msg': '添加组织失败：检测到其Alias与另一个组织重复' },
    'unKnownError': { 'rc': 4, 'msg': '添加组织失败：未知错误' }
};

var sendEmailToActMembers_rc = {
    'success': { 'rc': 0, 'msg': '发送邮件成功！' },
    'notAuth': { 'rc': 1, 'msg': '发送邮件失败：未授权' },
    'inputParaError': { 'rc': 2, 'msg': '发送邮件失败：输入参数错误' },
    'actNotFound': { 'rc': 3, 'msg': '发送邮件失败：没有找到指定的活动' },
    'actArchived': { 'rc': 4, 'msg': '发送邮件失败：指定的活动已归档，不能向队员发送信息' },
    'noValidMember': { 'rc': 5, 'msg': '发送邮件失败：提交的队员名单无效' },
    'unKnownError': { 'rc': 6, 'msg': '发送邮件失败：未知错误' }
};

var saveActBillStatement_rc = {
    'success': { 'rc': 0, 'msg': '保存活动账单成功！' },
    'notAuth': { 'rc': 1, 'msg': '保存活动账单失败：未授权' },
    'inputParaError': { 'rc': 2, 'msg': '保存活动账单失败：输入参数错误' },
    'actNotFound': { 'rc': 3, 'msg': '保存活动账单失败：没有找到指定的活动，或者活动已归档' },
    'unKnownError': { 'rc': 4, 'msg': '保存活动账单失败：未知错误' }
};

var saveActSummary_rc = {
    'success': { 'rc': 0, 'msg': '保存活动总结成功！' },
    'notAuth': { 'rc': 1, 'msg': '保存活动总结失败：未授权' },
    'inputParaError': { 'rc': 2, 'msg': '保存活动总结失败：输入参数错误' },
    'actNotFound': { 'rc': 3, 'msg': '保存活动总结失败：没有找到指定的活动，或者活动已归档' },
    'unKnownError': { 'rc': 4, 'msg': '保存活动总结失败：未知错误' }
};

var userInjection_rc = {
    'success': { 'rc': 0, 'msg': '用户注入成功！' },
    'notAuth': { 'rc': 1, 'msg': '用户注入失败：未授权' },
    'userNotFound': { 'rc': 2, 'msg': '用户注入失败：没有找到指定的用户，或无法注入' },
    'unKnownError': { 'rc': 3, 'msg': '用户注入失败：未知错误' }
};

var archiveOrgWeixin_rc = {
    'success': { 'rc': 0, 'msg': '删除微信自动回复成功！' },
    'notAuth': { 'rc': 1, 'msg': '删除微信自动回复失败：未授权' },
    'orgNotFound': { 'rc': 2, 'msg': '删除微信自动回复失败：当前组织无效或不可用' },
    'weixinNotFound': { 'rc': 3, 'msg': '删除微信自动回复失败：没有找到指定删除的微信回复条目' },
    'inputParaError': { 'rc': 4, 'msg': '删除微信自动回复失败：输入参数错误' },
    'unKnownError': { 'rc': 5, 'msg': '删除微信自动回复失败：未知错误' }
};

var submitOrgWeixinReply_rc = {
    'success': { 'rc': 0, 'msg': '保存微信自动回复成功！' },
    'notAuth': { 'rc': 1, 'msg': '保存微信自动回复失败：未授权' },
    'orgNotFound': { 'rc': 2, 'msg': '保存微信自动回复失败：当前组织无效或不可用' },
    'weixinNotFound': { 'rc': 3, 'msg': '保存微信自动回复失败：没有找到指定修改的微信回复条目' },
    'inputParaError': { 'rc': 4, 'msg': '保存微信自动回复失败：输入参数错误' },
    'imageUploadFail': { 'rc': 5, 'msg': '保存微信自动回复失败：图片上传失败，请联系系统管理员' },
    'unKnownError': { 'rc': 6, 'msg': '保存微信自动回复失败：未知错误' }
};

var setShareSummary_rc = {
    'success_share': { 'rc': 0, 'msg': '设置活动总结分享成功，感谢分享，同时你也可以看到其他社团分享的活动总结！' },
    'success_noshare': { 'rc': 1, 'msg': '取消活动总结分享成功' },
    'notAuth': { 'rc': 2, 'msg': '设置活动总结分享状态失败：未授权' },
    'orgNotFound': { 'rc': 3, 'msg': '设置活动总结分享状态失败：当前组织无效或不可用' },
    'alreadyShare': { 'rc': 4, 'msg': '设置活动总结分享状态失败：活动总结已经打开分享啦' },
    'alreadyNoShare': { 'rc': 5, 'msg': '设置活动总结分享状态失败：活动总结已经关闭分享啦' },
    'frequency': { 'rc': 6, 'msg': '设置活动总结分享状态失败：距离上一次修改时间小于一个月，无法再次修改' },
    'inputParaError': { 'rc': 7, 'msg': '设置活动总结分享状态失败：输入参数错误' },
    'unKnownError': { 'rc': 8, 'msg': '设置活动总结分享状态失败：未知错误' }
};

var setShareBillStatement_rc = {
    'success_share': { 'rc': 0, 'msg': '设置活动账目分享成功，感谢分享，同时你也可以看到其他社团分享的活动总结！' },
    'success_noshare': { 'rc': 1, 'msg': '取消活动账目分享成功' },
    'notAuth': { 'rc': 2, 'msg': '设置活动账目分享状态失败：未授权' },
    'orgNotFound': { 'rc': 3, 'msg': '设置活动账目分享状态失败：当前组织无效或不可用' },
    'alreadyShare': { 'rc': 4, 'msg': '设置活动账目分享状态失败：活动账目已经打开分享啦' },
    'alreadyNoShare': { 'rc': 5, 'msg': '设置活动账目分享状态失败：活动账目已经关闭分享啦' },
    'frequency': { 'rc': 6, 'msg': '设置活动账目分享状态失败：距离上一次修改时间小于一个月，无法再次修改' },
    'inputParaError': { 'rc': 7, 'msg': '设置活动账目分享状态失败：输入参数错误' },
    'unKnownError': { 'rc': 8, 'msg': '设置活动账目分享状态失败：未知错误' }
};

var submitNewForumPost_rc = {
    'success': { 'rc': 0, 'msg': '发帖成功！' },
    'notAuth': { 'rc': 1, 'msg': '发帖失败：未登录，请登录后再次尝试 ^^' },
    'userNotFound': { 'rc': 2, 'msg': '发帖失败：未找到当前用户' },
    'special': { 'rc': 3, 'msg': '发帖失败：特殊身份账户不允许发帖' },
    'inputParaError': { 'rc': 4, 'msg': '发帖失败：输入参数错误' },
    'imageUploadFail': { 'rc': 5, 'msg': '发帖失败：上传图片失败，请重试' },
    'frequency': { 'rc': 6, 'msg': '发帖失败：发帖过于频繁，请等待一段时间后再次尝试 ^^' },
    'shutup': { 'rc': 7, 'msg': '发帖失败：用户被禁言' },
    'unKnownError': { 'rc': 8, 'msg': '发帖失败：未知错误' },
    'actNotFound': { 'rc': 9, 'msg': '发帖失败：指定归属的活动未找到' },
    'orgNotFound': { 'rc': 10, 'msg': '发帖失败：指定归属的组织未找到' },
    'nickName': { 'rc': 11, 'msg': '发帖失败：请先设置论坛昵称，并再次尝试 ^^' },
    'options': { 'rc': 12, 'msg': '发帖失败：请为投票类型设置至少两个选项 ^^' }
};

var submitForumPostPrm_rc = {
    'success': { 'rc': 0, 'msg': '跟帖成功！' },
    'notAuth': { 'rc': 1, 'msg': '跟帖失败：未登录，请登录后再次尝试 ^^' },
    'userNotFound': { 'rc': 2, 'msg': '跟帖失败：未找到当前用户' },
    'special': { 'rc': 3, 'msg': '跟帖失败：特殊身份账户不允许跟帖' },
    'inputParaError': { 'rc': 4, 'msg': '跟帖失败：输入参数错误' },
    'imageUploadFail': { 'rc': 5, 'msg': '跟帖失败：上传图片失败，请重试' },
    'frequency': { 'rc': 6, 'msg': '跟帖失败：跟帖过于频繁，请等待一段时间后再次尝试 ^^' },
    'shutup': { 'rc': 7, 'msg': '跟帖失败：用户被禁言' },
    'unKnownError': { 'rc': 8, 'msg': '跟帖失败：未知错误' },
    'postPrmNotFound': { 'rc': 9, 'msg': '跟帖失败：未找到需要修改的帖子' },
    'postNotFound': { 'rc': 10, 'msg': '跟帖失败：未找到跟帖的目标贴' },
    'nickName': { 'rc': 11, 'msg': '跟帖失败：请先设置论坛昵称，并再次尝试 ^^' },
    'notArticle': { 'rc': 11, 'msg': '跟帖失败：编辑的跟帖是投票类型，无法编辑' }
};

var submitForumPostSec_rc = {
    'success': { 'rc': 0, 'msg': '回复成功！' },
    'notAuth': { 'rc': 1, 'msg': '回复失败：未登录，请登录后再次尝试 ^^' },
    'userNotFound': { 'rc': 2, 'msg': '回复失败：未找到当前用户' },
    'special': { 'rc': 3, 'msg': '回复失败：特殊身份账户不允许回复' },
    'inputParaError': { 'rc': 4, 'msg': '回复失败：输入参数错误' },
    'frequency': { 'rc': 5, 'msg': '回复失败：回复过于频繁，请等待一段时间后再次尝试 ^^' },
    'shutup': { 'rc': 6, 'msg': '回复失败：用户被禁言' },
    'unKnownError': { 'rc': 7, 'msg': '回复失败：未知错误' },
    'postPrmNotFound': { 'rc': 8, 'msg': '回复失败：未找到回复的目标贴' },
    'postNotFound': { 'rc': 9, 'msg': '回复失败：未找到回复的目标贴' },
    'nickName': { 'rc': 10, 'msg': '回复失败：请先设置论坛昵称，并再次尝试 ^^' },
    'replyToUserNotFound': { 'rc': 11, 'msg': '回复失败：没有找到你要回复的用户 TAT' }
};

var disactivateForumPost_rc = {
    'success': { 'rc': 0, 'msg': '删除帖子成功！' },
    'notAuth': { 'rc': 1, 'msg': '删除帖子失败：未授权' },
    'postNotFound': { 'rc': 2, 'msg': '删除帖子失败：未找到指定删除的帖子' },
    'inputParaError': { 'rc': 3, 'msg': '删除帖子失败：输入参数错误' },
    'unKnownError': { 'rc': 4, 'msg': '删除帖子失败：未知错误' },
    'disactive': { 'rc': 5, 'msg': '删除帖子失败：已删除' }
};

var disactivateForumPostPrm_rc = {
    'success': { 'rc': 0, 'msg': '删除跟帖成功！' },
    'notAuth': { 'rc': 1, 'msg': '删除跟帖失败：未授权' },
    'postPrmNotFound': { 'rc': 2, 'msg': '删除跟帖失败：未找到指定删除的跟帖' },
    'inputParaError': { 'rc': 3, 'msg': '删除跟帖失败：输入参数错误' },
    'unKnownError': { 'rc': 4, 'msg': '删除跟帖失败：未知错误' },
    'disactive': { 'rc': 5, 'msg': '删除跟帖失败：已删除' }
};

var disactivateForumPostSec_rc = {
    'success': { 'rc': 0, 'msg': '删除回复成功！' },
    'notAuth': { 'rc': 1, 'msg': '删除回复失败：未授权' },
    'postSecNotFound': { 'rc': 2, 'msg': '删除回复失败：未找到指定删除的回复' },
    'inputParaError': { 'rc': 3, 'msg': '删除回复失败：输入参数错误' },
    'unKnownError': { 'rc': 4, 'msg': '删除回复失败：未知错误' },
    'disactive': { 'rc': 5, 'msg': '删除回复失败：已删除' }
};

var activateForumPost_rc = {
    'success': { 'rc': 0, 'msg': '恢复帖子成功！' },
    'notAuth': { 'rc': 1, 'msg': '恢复帖子失败：未授权' },
    'postNotFound': { 'rc': 2, 'msg': '恢复帖子失败：未找到指定恢复的帖子' },
    'inputParaError': { 'rc': 3, 'msg': '恢复帖子失败：输入参数错误' },
    'unKnownError': { 'rc': 4, 'msg': '恢复帖子失败：未知错误' },
    'active': { 'rc': 5, 'msg': '恢复帖子失败：帖子并未删除' }
};

var activateForumPostPrm_rc = {
    'success': { 'rc': 0, 'msg': '恢复跟帖成功！' },
    'notAuth': { 'rc': 1, 'msg': '恢复跟帖失败：未授权' },
    'postPrmNotFound': { 'rc': 2, 'msg': '恢复跟帖失败：未找到指定恢复的跟帖' },
    'inputParaError': { 'rc': 3, 'msg': '恢复跟帖失败：输入参数错误' },
    'unKnownError': { 'rc': 4, 'msg': '恢复跟帖失败：未知错误' },
    'active': { 'rc': 5, 'msg': '恢复跟帖失败：跟帖并未删除' }
};

var activateForumPostSec_rc = {
    'success': { 'rc': 0, 'msg': '恢复回复成功！' },
    'notAuth': { 'rc': 1, 'msg': '恢复回复失败：未授权' },
    'postSecNotFound': { 'rc': 2, 'msg': '恢复回复失败：未找到指定恢复的回复' },
    'inputParaError': { 'rc': 3, 'msg': '恢复回复失败：输入参数错误' },
    'unKnownError': { 'rc': 4, 'msg': '恢复回复失败：未知错误' },
    'active': { 'rc': 5, 'msg': '恢复回复失败：回复并未删除' }
};

var emphasizeForumPost_rc = {
    'success': { 'rc': 0, 'msg': '帖子加精成功！' },
    'notAuth': { 'rc': 1, 'msg': '帖子加精失败：未授权' },
    'postNotFound': { 'rc': 2, 'msg': '帖子加精失败：未找到指定加精的帖子' },
    'inputParaError': { 'rc': 3, 'msg': '帖子加精失败：输入参数错误' },
    'unKnownError': { 'rc': 4, 'msg': '帖子加精失败：未知错误' },
    'emphasis': { 'rc': 5, 'msg': '帖子加精失败：原帖已加精' }
};

var fadeoutForumPost_rc = {
    'success': { 'rc': 0, 'msg': '帖子取消加精成功！' },
    'notAuth': { 'rc': 1, 'msg': '帖子取消加精失败：未授权' },
    'postNotFound': { 'rc': 2, 'msg': '帖子取消加精失败：未找到指定取消加精的帖子' },
    'inputParaError': { 'rc': 3, 'msg': '帖子取消加精失败：输入参数错误' },
    'unKnownError': { 'rc': 4, 'msg': '帖子取消加精失败：未知错误' },
    'noemphasis': { 'rc': 5, 'msg': '帖子取消加精失败：原帖并未加精' }
};

var updatePostTags_rc = {
    'success': { 'rc': 0, 'msg': '修改帖子标签成功！' },
    'notAuth': { 'rc': 1, 'msg': '修改帖子标签失败：未授权' },
    'postNotFound': { 'rc': 2, 'msg': '修改帖子标签失败：未找到指定的帖子' },
    'inputParaError': { 'rc': 3, 'msg': '修改帖子标签失败：输入参数错误' },
    'unKnownError': { 'rc': 4, 'msg': '修改帖子标签失败：未知错误' }
};

var updatePostLabel_rc = {
    'success': { 'rc': 0, 'msg': '修改帖子分类成功！' },
    'notAuth': { 'rc': 1, 'msg': '修改帖子分类失败：未授权' },
    'postNotFound': { 'rc': 2, 'msg': '修改帖子分类失败：未找到指定的帖子' },
    'inputParaError': { 'rc': 3, 'msg': '修改帖子分类失败：输入参数错误' },
    'unKnownError': { 'rc': 4, 'msg': '修改帖子分类失败：未知错误' }
};

var unauthorizeUserToPost_rc = {
    'success': { 'rc': 0, 'msg': '用户禁言成功！' },
    'notAuth': { 'rc': 1, 'msg': '用户禁言失败：未授权' },
    'orgNotFound': { 'rc': 2, 'msg': '用户禁言失败：未找到指定的组织' },
    'inputParaError': { 'rc': 3, 'msg': '用户禁言失败：输入参数错误' },
    'special': { 'rc': 4, 'msg': '用户禁言失败：特殊账户无法设置禁言' },
    'unKnownError': { 'rc': 5, 'msg': '用户禁言失败：未知错误' },
    'userNotFound': { 'rc': 6, 'msg': '用户禁言失败：未找到指定的用户' },
    'unauth': { 'rc': 6, 'msg': '用户禁言失败：指定用户已禁言' }
};

var authorizeUserToPost_rc = {
    'success': { 'rc': 0, 'msg': '用户解除禁言成功！' },
    'notAuth': { 'rc': 1, 'msg': '用户解除禁言失败：未授权' },
    'orgNotFound': { 'rc': 2, 'msg': '用户解除禁言失败：未找到指定的组织' },
    'inputParaError': { 'rc': 3, 'msg': '用户解除禁言失败：输入参数错误' },
    'special': { 'rc': 4, 'msg': '用户解除禁言失败：特殊账户无法设置' },
    'unKnownError': { 'rc': 5, 'msg': '用户解除禁言失败：未知错误' },
    'userNotFound': { 'rc': 6, 'msg': '用户解除禁言失败：未找到指定的用户' },
    'auth': { 'rc': 6, 'msg': '用户解除禁言失败：指定用户并未禁言' }
};

var voteForumPostPrm_rc = {
    'success': { 'rc': 0, 'msg': '投票成功！' },
    'notAuth': { 'rc': 1, 'msg': '投票失败：未登录，请登录后再次尝试 ^^' },
    'userNotFound': { 'rc': 2, 'msg': '投票失败：未找到当前用户' },
    'special': { 'rc': 3, 'msg': '投票失败：特殊身份账户不允许投票' },
    'inputParaError': { 'rc': 4, 'msg': '投票失败：输入参数错误' },
    'frequency': { 'rc': 5, 'msg': '投票失败：操作过于频繁，请等待一段时间后再次尝试 ^^' },
    'shutup': { 'rc': 6, 'msg': '投票失败：用户被禁言' },
    'unKnownError': { 'rc': 7, 'msg': '投票失败：未知错误' },
    'postPrmVoteNotFound': { 'rc': 8, 'msg': '投票失败：指定帖子未找到或者非投票类型' },
    'nickName': { 'rc': 9, 'msg': '投票失败：请先设置论坛昵称，并再次尝试 ^^' },
    'multi': { 'rc': 10, 'msg': '投票失败：请不要为单选类型的投票设置多个选项' },
    'voted': { 'rc': 11, 'msg': '投票失败：已投票，请不要重复尝试 ^^' },
    'postNotFound': { 'rc': 12, 'msg': '投票失败：指定帖子未找到或者非投票类型' },
    'option': { 'rc': 13, 'msg': '投票失败：选项不存在或不匹配' },
    'closed': { 'rc': 14, 'msg': '投票失败：来晚一步，投票已经关闭啦 TAT' }
};

var submitOrgWeixinId_rc = {
    'success': { 'rc': 0, 'msg': '更新微信公众号绑定成功！' },
    'inputParaError': { 'rc': 1, 'msg': '更新微信公众号绑定失败：输入参数错误' },
    'orgNotFound': { 'rc': 2, 'msg': '更新微信公众号绑定成功：组织未找到' },
    'notAuth': { 'rc': 3, 'msg': '更新微信公众号绑定成功：未授权' },
    'noChange': { 'rc': 4, 'msg': '更新微信公众号绑定成功：已绑定指定的公众号，无需更新' },
    'unKnownError': { 'rc': 5, 'msg': '更新微信公众号绑定成功：未知错误' },
};

var submitOrgWeixinKeyword_rc = {
    'success': { 'rc': 0, 'msg': '更新微信特殊关键词成功！' },
    'inputParaError': { 'rc': 1, 'msg': '更新微信特殊关键词失败：输入参数错误' },
    'orgNotFound': { 'rc': 2, 'msg': '更新微信特殊关键词失败：组织未找到' },
    'notAuth': { 'rc': 3, 'msg': '更新微信特殊关键词失败：未授权' },
    'invalidKeyword': { 'rc': 4, 'msg': '更新微信特殊关键词失败：此关键词无效，请联系系统管理员' },
    'noChange': { 'rc': 5, 'msg': '更新微信特殊关键词失败：无修改？' },
    'duplicate': { 'rc': 6, 'msg': '更新微信特殊关键词失败：此关键词和另一个特殊关键词重复，请换一个关键词' },
    'unKnownError': { 'rc': 7, 'msg': '更新微信特殊关键词失败：未知错误' },
};

var emailRegex = new RegExp('^(\\w)+(\\.\\w+)*@(\\w)+((\\.\\w+)+)$');
var phoneRegex = new RegExp('^[0-9]{11}$');
var md5Regex = new RegExp('^[a-fA-F0-9]{32}$');
var dateRegex = new RegExp('^[0-9]{1,4}/(((0?[1-9]|1[0-2])/(0?[1-9]|[12][0-9]))|((0?[13-9]|1[0-2])/30)|((0?[13578]|1[02])/31))$')

var forumPostLabels = [
    { 'name': '原创(YC)', 'code': 'yc' },
    { 'name': '闲话', 'code': 'xh' },
    { 'name': '求助', 'code': 'qz' },
    { 'name': '攻略', 'code': 'gl' },
    { 'name': '建议', 'code': 'jy' },
    { 'name': '投票', 'code': 'tp' }
];
var isForumPostLabel4Vote = function (label) {
    return label === '投票' || label === 'tp';
};

var translateForumPostLabel = function (labelCode) {
    var fl = forumPostLabels.filter(function (v) { return v.code === labelCode; });
    if (fl.length > 0)
        return fl[0];
    else
        return false;
};

exports.orgStatus = orgStatus;
exports.activityStatus = activityStatus;
exports.userStatus = userStatus;
exports.actMemberStatus = actMemberStatus;
exports.actMemberStatusNames = actMemberStatusNames;
exports.activeActMemberStatus = activeActMemberStatus;
exports.inactiveActMemberStatus = inactiveActMemberStatus;
exports.organizerUpdateMemberStatus = organizerUpdateMemberStatus;
exports.emailStatus = emailStatus;
exports.activeActStatus = activeActStatus;
exports.historicalActStatus = historicalActStatus;
exports.weixinReplyStatus = weixinReplyStatus;
exports.forumPostStatus = forumPostStatus;
exports.forumPostPrmStatus = forumPostPrmStatus;
exports.forumPostSecStatus = forumPostSecStatus;

exports.logLevel = logLevel;

exports.register_rc = register_rc;
exports.changePwd_rc = changePwd_rc;
exports.updateUserEmail_rc = updateUserEmail_rc;
exports.sendUserMessage_rc = sendUserMessage_rc;
exports.updateOrgPic_rc = updateOrgPic_rc;
exports.claimPwdReset_rc = claimPwdReset_rc;
exports.pwdResetRequest_rc = pwdResetRequest_rc;
exports.upsertOrgSignupSheetTemplate_rc = upsertOrgSignupSheetTemplate_rc;
exports.renameOrgSignupSheetTemplate_rc = renameOrgSignupSheetTemplate_rc;
exports.deleteOrgSignupSheetTemplate_rc = deleteOrgSignupSheetTemplate_rc;
exports.submitOrgActivity_rc = submitOrgActivity_rc;
exports.updateUserPersonalInfo_rc = updateUserPersonalInfo_rc;
exports.signUp_rc = signUp_rc;
exports.userActQuit_rc = userActQuit_rc;
exports.updateUserSignupSheet_rc = updateUserSignupSheet_rc;
exports.grantUserOrgAdmin_rc = grantUserOrgAdmin_rc;
exports.setOrgActArchived_rc = setOrgActArchived_rc;
exports.setOrgActRemoved_rc = setOrgActRemoved_rc;
exports.removeOrgAdmin_rc = removeOrgAdmin_rc;
exports.updateOrgIntro_rc = updateOrgIntro_rc;
exports.updateMemberStatus_rc = updateMemberStatus_rc;
exports.reactivateOrg_rc = reactivateOrg_rc;
exports.disactivateOrg_rc = disactivateOrg_rc;
exports.updateOrgBasicInfo_rc = updateOrgBasicInfo_rc;
exports.createNewOrg_rc = createNewOrg_rc;
exports.sendEmailToActMembers_rc = sendEmailToActMembers_rc;
exports.saveActBillStatement_rc = saveActBillStatement_rc;
exports.saveActSummary_rc = saveActSummary_rc;
exports.userInjection_rc = userInjection_rc;
exports.archiveOrgWeixin_rc = archiveOrgWeixin_rc;
exports.submitOrgWeixinReply_rc = submitOrgWeixinReply_rc;
exports.setShareSummary_rc = setShareSummary_rc;
exports.setShareBillStatement_rc = setShareBillStatement_rc;
exports.submitNewForumPost_rc = submitNewForumPost_rc;
exports.submitForumPostPrm_rc = submitForumPostPrm_rc;
exports.submitForumPostSec_rc = submitForumPostSec_rc;
exports.disactivateForumPost_rc = disactivateForumPost_rc;
exports.disactivateForumPostPrm_rc = disactivateForumPostPrm_rc;
exports.disactivateForumPostSec_rc = disactivateForumPostSec_rc;
exports.activateForumPost_rc = activateForumPost_rc;
exports.activateForumPostPrm_rc = activateForumPostPrm_rc;
exports.activateForumPostSec_rc = activateForumPostSec_rc;
exports.emphasizeForumPost_rc = emphasizeForumPost_rc;
exports.fadeoutForumPost_rc = fadeoutForumPost_rc;
exports.updatePostTags_rc = updatePostTags_rc;
exports.updatePostLabel_rc = updatePostLabel_rc;
exports.unauthorizeUserToPost_rc = unauthorizeUserToPost_rc;
exports.authorizeUserToPost_rc = authorizeUserToPost_rc;
exports.voteForumPostPrm_rc = voteForumPostPrm_rc;
exports.submitOrgWeixinId_rc = submitOrgWeixinId_rc;
exports.submitOrgWeixinKeyword_rc = submitOrgWeixinKeyword_rc;

exports.emailRegex = emailRegex;
exports.phoneRegex = phoneRegex;
exports.md5Regex = md5Regex;
exports.dateRegex = dateRegex;

exports.defaultPwd = defaultPwd;

exports.forumPostLabels = forumPostLabels;
exports.translateForumPostLabel = translateForumPostLabel;
exports.isForumPostLabel4Vote = isForumPostLabel4Vote;
